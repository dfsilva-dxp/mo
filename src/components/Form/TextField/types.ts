import { InputHTMLAttributes, ReactNode } from "react";

export type TextFieldProps = {
  label?: string;
  labelFor?: string;
  icon?: ReactNode;
  initialValue?: string;
  iconPosition?: "left" | "right";
  disabled?: boolean;
  onInput?: (value: string) => void;
} & InputHTMLAttributes<HTMLInputElement>;
