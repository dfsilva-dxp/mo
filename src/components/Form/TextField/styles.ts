import styled, { css, DefaultTheme } from "styled-components";

import { TextFieldProps } from "./types";

const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Input},
    ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.gray750};
      user-select: none;

      &::placeholder {
        color: currentColor;
      }
    }
  `
};

export const Wrapper = styled.div<Pick<TextFieldProps, "disabled">>`
  ${({ theme, disabled }) => css`
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`;

export const Group = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    gap: ${theme.spacings.xsmall};
    background: ${theme.colors.panel};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.1rem solid;
    border-color: ${theme.colors.panel};

    &:focus-within {
      border-color: ${theme.colors.primary};
    }
    &:focus-within ${Icon} {
      color: ${theme.colors.primary};
    }
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    cursor: pointer;
    font-weight: ${theme.font.medium};
  `}
`;

export const Input = styled.input`
  ${({ theme }) => css`
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;

    &::placeholder {
      color: ${theme.colors.gray500};
    }
  `}
`;

export const Icon = styled.div<Pick<TextFieldProps, "iconPosition">>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    font-size: 2.4rem;
    color: ${theme.colors.gray500};
    order: ${iconPosition === "right" ? 1 : 0};

    & > svg {
      width: 100%;
    }
  `}
`;
