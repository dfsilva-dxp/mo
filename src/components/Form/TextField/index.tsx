import { TextFieldProps } from "./types";

import { Group, Icon, Input, Label, Wrapper } from "./styles";
import { ChangeEvent, useState } from "react";

export default function TextField({
  label = "",
  labelFor = "",
  icon,
  initialValue = "",
  iconPosition,
  disabled = false,
  onInput,
  ...props
}: TextFieldProps) {
  const [value, setValue] = useState(initialValue);

  const onChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { value } = e.currentTarget;
    setValue(value);

    !!onInput && onInput(value);
  };

  return (
    <Wrapper disabled={disabled}>
      {!!label && <Label htmlFor={labelFor}>{label}</Label>}
      <Group>
        {!!icon && <Icon iconPosition={iconPosition}>{icon}</Icon>}
        <Input
          type="text"
          {...props}
          value={value}
          onChange={onChange}
          disabled={disabled}
        />
      </Group>
    </Wrapper>
  );
}
