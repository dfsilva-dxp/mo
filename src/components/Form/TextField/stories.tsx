import { Story, Meta } from "@storybook/react/types-6-0";
import { ComponentProps } from "react";
import { TiMail } from "react-icons/ti";

import TextField from ".";

export default {
  title: "Form/TextField",
  component: TextField,
  argTypes: {
    onInput: { action: "input" }
  },
  args: {
    id: "email",
    placeholder: "john.doe@gmail.com",
    disabled: false
  },
  parameters: {
    layout: "centered"
  }
} as Meta;

export const Basic: Story<ComponentProps<typeof TextField>> = ({
  ...props
}) => <TextField {...props} />;

export const WithoutIcon: Story<ComponentProps<typeof TextField>> = ({
  ...props
}) => <TextField {...props} />;

WithoutIcon.argTypes = {
  iconPosition: {
    options: ["left", "right"],
    control: { type: "select" }
  }
};

WithoutIcon.args = {
  label: "Without Icon",
  labelFor: "withoutIcon",
  id: "withoutIcon",
  placeholder: "Your placeholder"
};

export const WithIcon: Story<ComponentProps<typeof TextField>> = ({
  ...props
}) => <TextField {...props} icon={<TiMail />} />;

WithIcon.argTypes = {
  iconPosition: {
    options: ["left", "right"],
    control: { type: "select" }
  }
};

WithIcon.args = {
  iconPosition: "left",
  label: "With Icon",
  labelFor: "withIcon",
  id: "withIcon",
  placeholder: "Your placeholder"
};
