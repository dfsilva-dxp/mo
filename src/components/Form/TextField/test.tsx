import { screen, waitFor } from "@testing-library/react";
import { TiMail } from "react-icons/ti";
import userEvent from "@testing-library/user-event";

import { renderWithTheme } from "utils/tests/helpers";

import TextField from ".";

describe("TextField Component", () => {
  it("should renders with label", () => {
    const { container } = renderWithTheme(
      <TextField label="Label" labelFor="field" id="field" />
    );

    expect(screen.getByLabelText("Label")).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should renders without label", () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByLabelText("Label")).not.toBeInTheDocument();
  });

  it("should renders with placeholder", () => {
    renderWithTheme(
      <TextField
        label="Label"
        labelFor="field"
        id="field"
        placeholder="Your placeholder"
      />
    );

    expect(
      screen.getByPlaceholderText(/Your placeholder/i)
    ).toBeInTheDocument();
  });

  it("should renders an icon", () => {
    renderWithTheme(<TextField icon={<TiMail data-testid="icon" />} />);

    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });

  it("should renders an icon on the right side", () => {
    renderWithTheme(
      <TextField icon={<TiMail data-testid="icon" />} iconPosition="right" />
    );

    expect(screen.getByTestId("icon").parentElement).toHaveStyle({
      order: 1
    });
  });

  it("should renders without an icon", () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByTestId("icon")).not.toBeInTheDocument();
  });

  it("should change its value when typing", async () => {
    const onInput = jest.fn();

    renderWithTheme(
      <TextField onInput={onInput} label="Label" labelFor="field" id="field" />
    );

    const input = screen.getByRole("textbox");
    const text = "This is my new value!";

    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onInput).toHaveBeenCalledTimes(text.length);
    });

    expect(onInput).toHaveBeenCalledWith(text);
  });

  it("should is accessible by tab", () => {
    renderWithTheme(<TextField label="Label" labelFor="field" id="field" />);

    const input = screen.getByRole("textbox");

    expect(document.body).toHaveFocus();
    userEvent.tab();

    expect(input).toHaveFocus();
  });

  it("should does not changes its value when disabled", async () => {
    const onInput = jest.fn();
    renderWithTheme(
      <TextField
        onInput={onInput}
        label="Label"
        labelFor="field"
        id="field"
        disabled
      />
    );

    const input = screen.getByRole("textbox");
    expect(input).toBeDisabled();

    const text = "This is my new text";
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).not.toHaveValue(text);
    });
    expect(onInput).not.toHaveBeenCalled();
  });

  it("should is not accessible by tab when disabled", () => {
    renderWithTheme(
      <TextField label="Label" labelFor="field" id="field" disabled />
    );

    const input = screen.getByLabelText("Label");
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).not.toHaveFocus();
  });
});
