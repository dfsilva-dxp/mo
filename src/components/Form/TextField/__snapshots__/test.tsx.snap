// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Component should renders with label 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1.6rem;
  background: #1F2029;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.1rem solid;
  border-color: #1F2029;
}

.c1:focus-within {
  border-color: #FFC727;
}

.c0 {
  font-size: 1.4rem;
  cursor: pointer;
  font-weight: 500;
}

.c2 {
  font-family: default:Quicksand,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  heading: Roboto,-apple-system,BlinkMacSystemFont,'Segoe UI',Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c2::-webkit-input-placeholder {
  color: #616480;
}

.c2::-moz-placeholder {
  color: #616480;
}

.c2:-ms-input-placeholder {
  color: #616480;
}

.c2::placeholder {
  color: #616480;
}

<div
  class=""
>
  <label
    class="c0"
    for="field"
  >
    Label
  </label>
  <div
    class="c1"
  >
    <input
      class="c2"
      id="field"
      type="text"
      value=""
    />
  </div>
</div>
`;
